File: src/opendss_designer/core/phasing.py
"""Which phases an element actually connects to.

OpenDSS writes a connection as `bus.node.node` — `.2` is phase B, `.1.2` is
A and B, `.0` is neutral or ground. The numbers are positional and easy to get
wrong by hand, so the editor stores the engineer's own word for it: a `phasing`
param holding letters, "A", "BC", "ABC". This module is the only place the two
spellings meet.

Anything the letters cannot say — a neutral node, a centre tap, the fourth
node of a 4-wire delta — stays in the raw `busNodes`/`nodes1`/`nodes2` params
the importer already preserved. Those are rare, and the phase picker in the UI
declines to touch them rather than quietly dropping a node.
"""
from __future__ import annotations

import re

PHASE_LETTERS = "ABC"
LETTER_TO_NODE = {"A": 1, "B": 2, "C": 3}
NODE_TO_LETTER = {1: "A", 2: "B", 3: "C"}

_SUFFIX_RE = re.compile(r"^(\.\d+)+$")


def parse_phasing(raw: object) -> str | None:
    """Normalise a user-supplied phasing to unique upper-case letters, in ABC
    order. Anything unrecognisable returns None, which means "use the default
    for the phase count" — the behaviour every circuit had before pinning."""
    if not isinstance(raw, str):
        return None
    # Separators are courtesy ("A-B", "a, c"); every remaining character has
    # to be a phase. Picking the letters out of arbitrary text instead would
    # turn a typo into a confident, wrong pin.
    seen = [c for c in raw.upper() if c not in " -,._/+"]
    if not seen or len(set(seen)) != len(seen):
        return None
    if any(c not in PHASE_LETTERS for c in seen):
        return None
    # Sorted so ".2.1" and "BA" cannot produce two spellings of one connection;
    # OpenDSS cares about node order only for phase rotation, which the editor
    # does not model.
    return "".join(sorted(set(seen), key=PHASE_LETTERS.index))


def nodes_for(phasing: str | None) -> str:
    """Letters to the node suffix OpenDSS wants: "B" -> ".2"."""
    parsed = parse_phasing(phasing)
    if not parsed:
        return ""
    return "".join(f".{LETTER_TO_NODE[c]}" for c in parsed)


def phasing_from_suffix(suffix: object) -> str | None:
    """The inverse, for import: ".1.2" -> "AB". Returns None when the suffix
    says something the letters cannot — a neutral, a repeat, a node past 3 —
    so the caller keeps the raw text instead."""
    if not isinstance(suffix, str) or not _SUFFIX_RE.match(suffix):
        return None
    nodes = [int(part) for part in suffix.split(".") if part]
    if not nodes or len(set(nodes)) != len(nodes):
        return None
    if any(n not in NODE_TO_LETTER for n in nodes):
        return None
    return "".join(sorted((NODE_TO_LETTER[n] for n in nodes), key=PHASE_LETTERS.index))


def phase_count(params: dict, default: int = 3) -> int:
    """The element's declared phase count, clamped to 1-3.

    The clamp is not cosmetic: a huge value builds a pathological command, so
    this is the one place it is enforced for both the compiler and the
    validator.
    """
    raw = params.get("phases", default)
    try:
        n = int(float(raw))  # type: ignore[arg-type]
    except (TypeError, ValueError):
        return default
    return min(max(n, 1), 3)


def default_phasing(phases: int) -> str:
    """What OpenDSS assumes when no suffix is given: the first N phases."""
    return PHASE_LETTERS[: max(1, min(3, int(phases)))]


def phase_set(phasing: str | None, phases: int) -> frozenset[str]:
    """The phases an element occupies, pinned or defaulted."""
    return frozenset(parse_phasing(phasing) or default_phasing(phases))

File: frontend/src/lib/phasing.ts
/** Which of the three phases an element connects to.
 *
 *  Mirror of `core/phasing.py`. The circuit stores letters ("B", "AC"); the
 *  compiler turns them into the node suffixes OpenDSS wants. An element with
 *  no `phasing` takes the first N phases, which is what OpenDSS assumes when a
 *  connection names no nodes -- so a drawing made before pinning existed still
 *  means exactly what it meant then.
 */
import type { BusResult, Params } from '../types/circuit'
import { NEUTRAL } from './colorScale'

export const PHASE_LETTERS = 'ABC'

/** Unique A/B/C letters in ABC order, or null if the text says nothing usable. */
export function parsePhasing(raw: unknown): string | null {
  if (typeof raw !== 'string') return null
  // Separators are courtesy ("A-B", "a, c"); every remaining character has to
  // be a phase. Picking the letters out of arbitrary text instead would turn
  // a typo into a confident, wrong pin.
  const seen = [...raw.toUpperCase()].filter((c) => !' -,._/+'.includes(c))
  if (!seen.length || new Set(seen).size !== seen.length) return null
  if (seen.some((c) => !PHASE_LETTERS.includes(c))) return null
  return [...new Set(seen)].sort((a, b) => PHASE_LETTERS.indexOf(a) - PHASE_LETTERS.indexOf(b)).join('')
}

export function phaseCount(params: Params | undefined): number {
  const n = Number(params?.phases)
  if (!Number.isFinite(n)) return 3
  return Math.min(Math.max(Math.trunc(n), 1), 3)
}

/** What the element actually connects to: the pin, or the default for its
 *  phase count. */
export function effectivePhasing(params: Params | undefined): string {
  return parsePhasing(params?.phasing) ?? PHASE_LETTERS.slice(0, phaseCount(params))
}

/** A label for the one-line, or null when there is nothing worth saying --
 *  a plain 3-phase element is the common case and labelling it every time
 *  would bury the laterals that do carry information. */
export function phaseLabel(params: Params | undefined): string | null {
  const phasing = effectivePhasing(params)
  return phasing === PHASE_LETTERS ? null : phasing.split('').join('-')
}

/** The letter for an OpenDSS node number: 1 -> A, 2 -> B, 3 -> C. The neutral
 *  (0) is N; anything else is left as the number it came with, since the
 *  letters have nothing to say about it. */
export function nodeLetter(node: number | undefined): string {
  if (node == null) return '?'
  if (node === 0) return 'N'
  return PHASE_LETTERS[node - 1] ?? String(node)
}

/** Colours for the 'phases' overlay. There is no universal convention --
 *  every utility maps its own GIS colours -- so this is one constant to hand
 *  to a theme later, not a preference now. A three-phase element keeps the
 *  plain ink so a healthy trunk stays quiet, and a bus nothing reaches gets
 *  the de-energised grey the voltage overlay already uses. */
export const PHASE_COLORS: Record<string, string> = { A: '#c62828', B: '#1565c0', C: '#2e7d32' }
export const TWO_PHASE_COLOR = '#6a1b9a'
export const THREE_PHASE_COLOR = '#263238'
export const NO_PHASE_COLOR = NEUTRAL

export const PHASE_LEGEND: { label: string; color: string }[] = [
  { label: 'A', color: PHASE_COLORS.A },
  { label: 'B', color: PHASE_COLORS.B },
  { label: 'C', color: PHASE_COLORS.C },
  { label: 'Two-phase', color: TWO_PHASE_COLOR },
  { label: 'Three-phase', color: THREE_PHASE_COLOR },
  { label: 'Unfed', color: NO_PHASE_COLOR },
]

/** The stroke for a set of phase letters. `undefined`/`null` is "not known
 *  yet" (validation has not answered) and draws as plain ink; "" is "known
 *  to be nothing", which is the case worth seeing. */
export function phaseColor(letters: string | null | undefined): string {
  if (letters == null) return THREE_PHASE_COLOR
  if (letters === '') return NO_PHASE_COLOR
  const parsed = parsePhasing(letters)
  if (!parsed || parsed.length === 3) return THREE_PHASE_COLOR
  if (parsed.length === 2) return TWO_PHASE_COLOR
  return PHASE_COLORS[parsed]
}

/** How far apart the phases of a bus may sit before it counts as unbalanced
 *  on the drawing. Below this a bus is reported as one voltage. */
export const BALANCE_TOLERANCE_PU = 0.002

/** The phase that holds the bus's lowest voltage, or null when naming it
 *  would add nothing: a three-phase bus whose phases agree within the
 *  tolerance, or a step with no per-phase data (time series). A bus short of
 *  three phases always names it, since the letter then says which phase the
 *  lateral is on. */
export function weakestPhase(bus: BusResult): string | null {
  const phases = bus.nodes.filter((n) => n >= 1 && n <= 3)
  if (!phases.length || bus.vmagPu.length !== bus.nodes.length) return null
  let minIdx = -1
  let minV = Infinity
  let maxV = -Infinity
  bus.nodes.forEach((n, i) => {
    if (n < 1 || n > 3) return
    const v = bus.vmagPu[i]
    if (v < minV) {
      minV = v
      minIdx = i
    }
    if (v > maxV) maxV = v
  })
  if (phases.length === 3 && maxV - minV <= BALANCE_TOLERANCE_PU) return null
  return nodeLetter(bus.nodes[minIdx])
}
